model_compression_toolkit/core/keras/hessian/weights_trace_hessian_calculator_keras.py

```diff
@@ -50,5 +53,112 @@ def __init__(self,
+        according to the requested granularity. If for example the node is Conv2D with a kernel
+        shape of (3, 3, 3, 2) (namely, 3 input channels, 2 output channels and kernel size of 3x3)
+        and the required granularity is HessianInfoGranularity.PER_TENSOR the result shape will be (1,),
+        for HessianInfoGranularity.PER_OUTPUT_CHANNEL the shape will be (2,) and for
+        HessianInfoGranularity.PER_ELEMENT a shape of (3, 3, 3, 2).
+
+        Returns:  The computed scores as numpy ndarray.
+
+        """
+        # Check if the target node's layer type is supported
+        if self.hessian_request.target_node.layer_class not in [Conv2D, Dense, Conv2DTranspose, DepthwiseConv2D]:
+            Logger.error(
+                f"{self.hessian_request.target_node.type} is not supported for Hessian info w.r.t weights.")
+
+        # Construct the Keras float model for inference
+        model, _ = FloatKerasModelBuilder(graph=self.graph).build_model()
+        # TODO: what to do in reuse
+
+        # Get the weight attributes for the target node type
+        weight_attributes = DEFAULT_KERAS_INFO.get_kernel_op_attributes(self.hessian_request.target_node.type)
+        assert len(weight_attributes) == 1
+
+        # Get the weight tensor for the target node
+        weight_tensor = getattr(model.get_layer(self.hessian_request.target_node.name), weight_attributes[0])
+
+        # Get the output channel index (needed for HessianInfoGranularity.PER_OUTPUT_CHANNEL case)
+        output_channel_axis, _ = DEFAULT_KERAS_INFO.kernel_channels_mapping.get(
+            self.hessian_request.target_node.type)
+
+        # Initiate a gradient tape for automatic differentiation
+        with tf.GradientTape(persistent=True) as tape:
+            # Perform a forward pass (inference) to get the output, while watching
+            # the input tensor for gradient computation
+            tape.watch(self.input_images)
+            outputs = model(self.input_images)
+
+            # Combine outputs if the model returns multiple output tensors
+            output = self._concat_outputs(outputs)
+
+            approximation_per_iteration = []
+            for j in range(self.num_iterations_for_approximation):  # Approximation iterations
+                # Getting a random vector with normal distribution and the same shape as the model output
+                v = tf.random.normal(shape=output.shape)
+                f_v = tf.reduce_sum(v * output)
+
+                # Stop recording operations for automatic differentiation
+                with tape.stop_recording():
+                    # Compute gradients of f_v with respect to the weights
+                    gradients = tape.gradient(f_v, weight_tensor)
+
+                    # Reshape the gradients based on the granularity (whole tensor, per channel, or per element)
+                    num_of_scores = self._get_num_scores_by_granularity(gradients,
```

Comment:
this function should be above the grad calculation and gets weights and not gradiets. You don't need to calc it each iteration.